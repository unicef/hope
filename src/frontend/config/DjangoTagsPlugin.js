const path = require('path');
const fs = require('fs-extra');
const paths = require("./paths");

function copyPublicFolder() {
  console.log("Copying public folder", paths.appPublic, paths.appBuild)
  fs.copySync(paths.appPublic, paths.appBuild, {
    dereference: true,
    filter: file => file !== paths.appHtml,
  });
}
class DjangoTagsPlugin {
  constructor(options) {
    this.options = options;
  }

  apply(compiler) {
    compiler.hooks.done.tapAsync('DjangoTagsPlugin', (stats, callback) => {
      const tagToSearch = compiler.options.output.publicPath;
      const outputPath = compiler.options.output.path;
      const indexPath = path.join(outputPath, 'index.html');
      const targetPath = path.join(
        this.options.djangoTemplatePath,
        'index.html',
      );

      // Read index.html
      fs.readFile(indexPath, 'utf8', (err, data) => {
        if (err) {
          console.error(err);
          return callback();
        }


        const modifiedData = data.replaceAll(new RegExp(tagToSearch+"\\/(.*?)\"","g"), `{% static "/${this.options.djangoAppName}/$1" %}"`);
        console.log("Modified index.html");
        fs.writeFile(targetPath, modifiedData, 'utf8', (writeErr) => {
          if (writeErr) {
            console.error(writeErr);
            return callback();
          }

          copyPublicFolder()
          // Delete the original file
          fs.unlink(indexPath, (unlinkErr) => {
            if (unlinkErr) {
              console.error(unlinkErr);
            }
            callback();
          });
        });
      });
    });
  }
}



module.exports = DjangoTagsPlugin;
