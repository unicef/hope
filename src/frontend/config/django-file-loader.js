const path = require('path');

const {getOptions, interpolateName} = require('loader-utils');
const validateOptions = require('schema-utils');
const loaderUtils =  require('loader-utils');
const schema = require('./options.json');
const {normalizePath} = require('./utils');


function loader(content) {
  const options = loaderUtils.getOptions(this) || {};

  validateOptions(schema, options, {
    name: 'File Loader',
    baseDataPath: 'options',
  });

  const context = options.context || this.rootContext;

  const url = loaderUtils.interpolateName(
    this,
    options.name || '[contenthash].[ext]',
    {
      context,
      content,
      regExp: options.regExp,
    }
  );

  let outputPath = url;

  if (options.outputPath) {
    if (typeof options.outputPath === 'function') {
      outputPath = options.outputPath(url, this.resourcePath, context);
    } else {
      outputPath = path.posix.join(options.outputPath, url);
    }
  }

  let publicPath = `addPublicPath(${JSON.stringify(outputPath)})`;

  if (options.publicPath) {
    if (typeof options.publicPath === 'function') {
      publicPath = options.publicPath(url, this.resourcePath, context);
    } else {
      publicPath = `${
        options.publicPath.endsWith('/')
          ? options.publicPath
          : `${options.publicPath}/`
      }${url}`;
    }

    publicPath = JSON.stringify(publicPath);
  }

  if (options.postTransformPublicPath) {
    publicPath = options.postTransformPublicPath(publicPath);
  }

  if (typeof options.emitFile === 'undefined' || options.emitFile) {
    this.emitFile(outputPath, content);
  }

  return `${
    options.esModules ? 'export default' : 'module.exports ='
  } ${publicPath};`;
}
loader.raw = true;
module.exports = loader;