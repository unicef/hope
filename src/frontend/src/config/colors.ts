export const NAVY_BLUE = '#003C8F';
export const DARK_GRAY = '#233944';
