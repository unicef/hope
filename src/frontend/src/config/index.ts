export const API_BASE_URL = '/api';
export const GRAPHQL_URL = `${API_BASE_URL}/graphql`;
export const LOGIN_URL = '/api/login/azuread-tenant-oauth2/';
export const DATE_FORMAT = 'D MMM YYYY';
export const DATE_FORMAT_LONG = 'D MMM YYYY h:mm A';
export const AUTO_LOGOUT_MINUTES = 15;
export const AUTO_LOGOUT_MILLIS = 1000 * 60 * AUTO_LOGOUT_MINUTES;
