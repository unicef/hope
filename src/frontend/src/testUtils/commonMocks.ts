import { vi } from 'vitest';
import React from 'react';

/**
 * Common mock for @hooks/useBaseUrl hook
 */
export const mockUseBaseUrl = (
  businessArea = 'afghanistan',
  programId = 'test-program',
  baseUrl = 'afghanistan/test-program',
) => {
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({ businessArea, programId, baseUrl }),
  }));
};

/**
 * Common mock for useProgramContext hook
 */
export const mockUseProgramContext = (
  groupLabel = 'Household',
  groupLabelPlural = 'Households',
) => {
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel,
          groupLabelPlural,
        },
      },
    }),
  }));
};

/**
 * Common mock for react-router-dom
 */
export const mockReactRouterDom = () => {
  vi.mock('react-router-dom', () => ({
    useNavigate: () => vi.fn(),
    useLocation: () => ({
      pathname: '/test-path',
      search: '',
      hash: '',
      state: null,
      key: 'test-key',
    }),
    Link: ({ children, to, ...props }: any) =>
      React.createElement('a', { href: to, ...props }, children),
  }));
};
/**
 * Common mock for @utils/utils module
 */
export const utilsMock = {
  createApiParams: vi.fn((baseParams, queryParams, options) => ({
    ...baseParams,
    ...queryParams,
    ...options,
  })),
  adjustHeadCells: vi.fn((headCells) => headCells),
  formatCurrencyWithSymbol: vi.fn(
    (amount, currency) => `${amount} ${currency}`,
  ),
  householdStatusToColor: vi.fn(() => '#000000'),
  filterEmptyParams: vi.fn((params) => {
    if (!params) return {};
    return Object.fromEntries(
      Object.entries(params).filter(([, value]) => {
        return value !== undefined && value !== null && value !== '';
      }),
    );
  }),
  opacityToHex: vi.fn((opacity) =>
    Math.round(opacity * 255)
      .toString(16)
      .padStart(2, '0'),
  ),
  isPermissionDeniedError: vi.fn((error) => error?.response?.status === 403),
  dateToIsoString: vi.fn((date) => date?.toISOString?.() || date),
  choicesToDict: vi.fn((choices) => {
    if (!choices) return {};
    return choices.reduce((acc, choice) => {
      acc[choice.value] = choice.name;
      return acc;
    }, {});
  }),
  populationStatusToColor: vi.fn(() => 'primary'),
  sexToCapitalize: vi.fn(
    (sex) => sex?.charAt(0).toUpperCase() + sex?.slice(1).toLowerCase(),
  ),
  programStatusToColor: vi.fn(() => 'primary'),
  formatCurrency: vi.fn((amount) => `$${amount?.toLocaleString() || '0'}`),
  columnToOrderBy: vi.fn((column) => column || 'name'),
};

/**
 * Helper function to mock the @utils/utils module
 */
export const mockUtils = () => {
  vi.mock('@utils/utils', () => utilsMock);
};

/**
 * Common mock for RestService
 */
export const mockRestService = () => {
  vi.mock('@restgenerated/services/RestService', () => ({
    RestService: {
      restBusinessAreasProgramsHouseholdsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsCountRetrieve: vi.fn(),
      restBusinessAreasProgramsIndividualsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsMembersList: vi.fn(),
      restBusinessAreasProgramsList: vi.fn(),
      restBusinessAreasProgramsCountRetrieve: vi.fn(),
      restBusinessAreasUsersProfileRetrieve: vi.fn(),
      // Add other commonly used RestService methods as needed
    },
  }));
};

/**
 * Helper function to set up all common mocks at once
 * Usage in test files:
 *
 * import { setupCommonMocks } from 'src/testUtils/commonMocks';
 *
 * // At the top level of your test file
 * setupCommonMocks();
 */
export const setupCommonMocks = () => {
  // Mock useBaseUrl
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({
      businessArea: 'afghanistan',
      programId: 'test-program',
      baseUrl: 'afghanistan/test-program',
    }),
  }));

  // Mock useProgramContext
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel: 'Household',
          groupLabelPlural: 'Households',
        },
      },
    }),
  }));

  mockReactRouterDom();
  mockUtils();
  mockRestService();
};
