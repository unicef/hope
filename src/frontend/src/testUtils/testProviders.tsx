import { ThemeProvider } from '@mui/material';
import CssBaseline from '@mui/material/CssBaseline';
import * as React from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { theme } from '../theme';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';

interface TestProvidersProps {
  children: React.ReactNode;
}

export const TestProviders: React.FC<TestProvidersProps> = ({ children }) => (
  <ThemeProvider theme={theme}>
    <StyledThemeProvider theme={theme}>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <CssBaseline />
        {children}
      </LocalizationProvider>
    </StyledThemeProvider>
  </ThemeProvider>
);
