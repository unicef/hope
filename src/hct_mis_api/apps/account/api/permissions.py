from typing import Any

from rest_framework.permissions import BasePermission

from hct_mis_api.apps.account.permissions import check_permissions


class BaseRestPermission(BasePermission):
    """
    Base class for custom permissions.
    """

    def _get_permissions(self, view: Any) -> Any:
        if hasattr(view, "permissions_by_action"):
            if view.action in view.permissions_by_action:
                return view.permissions_by_action[view.action]
            elif view.action == "count":
                return view.permissions_by_action["list"]
        return view.PERMISSIONS

    def has_permission(self, request: Any, view: Any) -> bool:
        user = request.user
        permissions = self._get_permissions(view)
        kwargs = {
            "business_area": request.parser_context.get("kwargs", {}).get("business_area_slug"),
            "program": request.parser_context.get("kwargs", {}).get("program_slug"),
            "Program": request.parser_context.get("kwargs", {}).get("program_id"),  # TODO: GraphQL - remove at the end
        }

        return check_permissions(user, permissions, **kwargs)
