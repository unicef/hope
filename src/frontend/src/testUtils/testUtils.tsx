import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
