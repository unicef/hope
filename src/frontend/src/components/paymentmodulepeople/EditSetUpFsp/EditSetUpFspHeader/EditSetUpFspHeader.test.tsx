import * as React from 'react';
import { PERMISSIONS } from '../../../../config/permissions';
import { render } from '../../../../testUtils/testUtils';
import { EditSetUpFspHeader } from './EditSetUpFspHeader';

describe('components/paymentmodule/EditSetUpFsp/EditSetUpFspHeader', () => {
  it('should render', () => {
    const { container } = render(
      <EditSetUpFspHeader permissions={[PERMISSIONS.PM_VIEW_LIST]} />,
    );
    expect(container).toMatchSnapshot();
  });
});
