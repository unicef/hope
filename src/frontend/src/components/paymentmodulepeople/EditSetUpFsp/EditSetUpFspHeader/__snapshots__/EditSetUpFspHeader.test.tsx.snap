// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/paymentmodule/EditSetUpFsp/EditSetUpFspHeader should render 1`] = `
<div>
  <div
    class="css-1x0er1"
    data-cy="page-header-container"
  >
    <div
      class="css-n88ekr"
    >
      <div
        class="css-ozdmop"
      >
        <div>
          <div
            class="MuiBox-root css-70qvj9"
          >
            <div
              class="css-9dlhcl"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-16xl4zq-MuiTypography-root"
                data-cy="page-header-title"
              >
                Edit Set up FSP
              </h5>
            </div>
            <div
              class="MuiBox-root css-15aqakd"
              data-cy="button-admin"
            />
          </div>
        </div>
        <div
          class="css-yeouz0"
        />
      </div>
    </div>
    <div
      class="css-8hfbf4"
    />
  </div>
</div>
`;
